contracts/security-token/ERC865.sol

```diff
@@ -0,0 +1,103 @@
+pragma solidity ^0.4.18;
+
+import "./CheckpointToken.sol";
+
+/* Largely copied from https://github.com/OpenZeppelin/openzeppelin-solidity/pull/741/files */
+
+contract ERC865 is CheckpointToken {
+  mapping(bytes => bool) signatures;
+
+  event TransferPreSigned(address indexed from, address indexed to, address indexed delegate, uint256 amount, uint256 fee);
+  event Debug(address from, bytes32 hash);
+
+  /**
+    * @notice Submit a presigned transfer
+    * @param _signature bytes The signature, issued by the owner.
+    * @param _to address The address which you want to transfer to.
+    * @param _value uint256 The amount of tokens to be transferred.
+    * @param _fee uint256 The amount of tokens paid to msg.sender, by the owner.
+    * @param _nonce uint256 Presigned transaction number
+    */
+  function transferPreSigned(
+    bytes _signature,
+    address _to,
+    uint256 _value,
+    uint256 _fee,
+    uint256 _nonce
+  )
+    public
+    returns (bool)
+  {
+    require(_to != address(0));
+    require(signatures[_signature] == false);
+    bytes32 hashedTx = transferPreSignedHashing(address(this), _to, _value, _fee, _nonce);
+    address from = recover(hashedTx, _signature);
+    require(from != address(0));
+
+    transferInternal(from, _to, _value);
+    transferInternal(from, msg.sender, _fee);
+
+    signatures[_signature] = true;
+    TransferPreSigned(from, _to, msg.sender, _value, _fee);
+    return true;
+  }
+
+  /**
+    * @notice Hash (keccak256) of the payload used by transferPreSigned
+    * @param _token address The address of the token.
+    * @param _to address The address which you want to transfer to.
+    * @param _value uint256 The amount of tokens to be transferred.
+    * @param _fee uint256 The amount of tokens paid to msg.sender, by the owner.
+    * @param _nonce uint256 Presigned transaction number.
+    */
+  function transferPreSignedHashing(
+    address _token,
+    address _to,
+    uint256 _value,
+    uint256 _fee,
+    uint256 _nonce
+  )
+    public
+    pure
+    returns (bytes32)
+  {
+    /* "48664c16": transferPreSignedHashing(address,address,address,uint256,uint256,uint256) */
+    return keccak256(bytes4(0x48664c16), _token, _to, _value, _fee, _nonce);
+  }
+
+  /**
+    * @notice Recover signer address from a message by using his signature
+    * @param hash bytes32 message, the hash is the signed message. What is recovered is the signer address.
+    * @param sig bytes signature, the signature is generated using web3.eth.sign()
+    */
+  function recover(bytes32 hash, bytes sig) public pure returns (address) {
```

Comment:
Can you add a comment why we need a customer recover() and ecrecover() is not enough. It is not immediate clear from the context.

Reply:
Yes, comment added :slightly_smiling_face: